design/20210324.certificate-request-approve-deny.md

```diff
@@ -0,0 +1,191 @@
+---
+title: CertificateRequest Approve and Denied Conditions
+authors:
+  - "@joshvanl"
+editor: "@joshvanl"
+creation-date: 2021-02-03
+last-updated: 2021-02-03
+status: implementable
+---
+
+# Approved and Denied Conditions
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [API Behaviour](#api-behaviour)
+- [RBAC](#rbac)
+  * [Scope](#scope)
+  * [API Group](#api-group)
+  * [Resource](#resource)
+  * [Verbs](#verbs)
+  * [Resource Names](#resource-names)
+- [cert-manager Approver](#cert-manager-approver)
+- [Certificates Controller Changes](#certificates-controller-changes)
+  * [Issuing Controller](#issuing-controller)
+<!-- /toc -->
+
+## Summary
+
+The CertificateRequest resource has [Approved and Denied
+conditions](https://github.com/jetstack/cert-manager/blob/7204284063702358164713379b07b73642ea9bec/pkg/apis/certmanager/v1/types_certificaterequest.go#L194).
+These conditions are based upon the [certificates.k8s.io
+CertificateSigningRequest](https://github.com/kubernetes/api/blob/3a2d6b5bb7f84daa2125131865eb81b151448d5b/certificates/v1/types.go#L198)
+conditions of the same name. The purpose of these conditions is so that users,
+or "approvers", are able to mark a request as either Approved or Denied. This
+condition gates an Issuer from signing the request. An Issuer waits for a
+CertificateRequest to have an Approved condition before signing. A
+CertificateRequest with a Denied condition will never be signed.
+
+
+## API Behaviour
+
+The Approved and Denied conditions are two distinct condition types on the
+CertificateRequest. These conditions must always have the status of True, and
+are mutually exclusive (i.e. a CertificateRequest cannot have an Approved and
+Denied condition simultaneously). This behaviour is enforced in the cert-manager
+Webhook component.
+
+An "approver" is an entity that is responsible for setting the Approved/Denied
+conditions. It is up to the approver's implementation as to what
+CertificateRequests are managed by that approver.
+
+The Reason field of the Approved/Denied condition should be set to *who* set the
```

Comment:
I don't think this piece is true? ('who' I mean) - this can be arbitrary text. If we don't want to allow arbitrary text, we need to clearly define the format of *who* - but I don't personally think this is what we want

Reply:
I agree, we want to have this as arbitrary text. The intention with this section is to give approver implementers a guide to what to fill in this field with since "Reason" can mean different things. By having a guide, we will have better consistency and thus better UX.